### ISSUE TITLE
Using escape characters in variable strings

### ISSUE THREAD

**Issue by jamie1192** [flutter/flutter#37668]
While this could be more of a SO-type question, [I have posted there already to no useful response](https://stackoverflow.com/questions/57213902/new-line-n-not-being-respected-when-reading-from-sqflite-result) so I'm out of options and this is currently a blocker for many of our production apps that rely on data-driven icons and string formatting.

I'm trying to use escape characters in both SQFlite data and from other model class variables, neither of which seem to work at all.

For example, I have a FontAwesome .ttf that I'm trying to render fonts from by referencing the unicode characters  eg. `\uf535'. If I hardcode the text value like this:

```dart
RichText(
  text: "\uf535"
)
```
then the icon renders correctly, but doing:

```dart
class Model {
  String iconString = "\uf535";
}

RichText(
  text: "${model.iconString}"
)
``` 
renders the literal string, and no icon/escape formatting works at all. Is there some sort of trick to getting this to work or is something going funky somewhere?

**Comment by goderbauer** [comment]
@BondarenkoStas Who is the customer that's blocked by this? Can you please add the specific customer label for that customer? Thanks!

**Comment by goderbauer** [comment]
@a-siva Does dart handle those two strings differently?

/cc @GaryQian for text rendering

**Comment by goderbauer** [comment]
For what it is worth, I cannot reproduce the behavior. The code below prints "Hello 😆😆" just as expected.

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class Model {
  String foo = '\u{1F606}';
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    Model model = Model();
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: Scaffold(
        body: Center(
          child: RichText(
            text: TextSpan(
              style: TextStyle(color: Colors.black),
              text: 'Hello \u{1F606} ${model.foo}',
            ),
          ),
        ),
      ),
    );
  }
}
```

Can you post a complete minimal code example that we can copy and run to reproduce the problem?

**Comment by jamie1192** [comment]
Thanks for the response @goderbauer .

So a little while after I posted this I found that if the string is hard-coded to a variable like in your example above, it works fine as you've just shown.

But for our apps that are heavily data-driven, we need to pull content from an SQLite database and there's many columns that require line breaks and FontAwesome unicode values in the string for questionnaire and response data, so this is when the issue occurs. When the app retrieves the SQLite results, it returns the literal string and both line breaking/FontAwesome values etc aren't rendered as they should be ([see screenshot here for debug output](https://cdn.discordapp.com/attachments/421445334716121091/608196724040925188/Screen_Shot_2019-08-06_at_5.16.17_pm.png) and [my Stack Overflow post regarding this here](https://stackoverflow.com/questions/57213902/new-line-n-not-being-respected-when-reading-from-sqflite-result)).

Also for context, the app is using the flutter_bloc and sqflite packages for the above functionality so I'm not sure if the issue could be due to either of those (though I have asked around to both devs and they said their packages weren't responsible for this).

Any ideas?

**Comment by jamie1192** [comment]
here's a quick sample project to demonstrate - https://github.com/jamie1192/EscapeStringSample

**Comment by BondarenkoStas** [comment]
sorry @goderbauer, I used the label in a wrong way.

**Comment by MuhammadIdees** [comment]
Hello, I was having somewhat the same issue with my app. The problem is not with flutter it's the way data is being stored in the database. I was trying to store some Unicode escape characters but the character '\u' is taken as '\\u' (or '\' + 'u').

So what I did was after retrieving data from SQL I changed the \\u to \u. I wish it were easy as replaceAll('\\u', '\u') but it's not. I have a python implementation of that if you're interested I can share it with you but you have to implement something similar in your dart file.

**Comment by jamie1192** [comment]
@MuhammadIdees yeah I do have a replaceAll function that I was suing to get around this but was hoping there was a more ideal and efficient way to get around this without having to run that replace on essentially every single column before binding to the view :/

**Comment by nxhiep** [comment]
> While this could be more of a SO-type question, [I have posted there already to no useful response](https://stackoverflow.com/questions/57213902/new-line-n-not-being-respected-when-reading-from-sqflite-result) so I'm out of options and this is currently a blocker for many of our production apps that rely on data-driven icons and string formatting.
> 
> I'm trying to use escape characters in both SQFlite data and from other model class variables, neither of which seem to work at all.
> 
> For example, I have a FontAwesome .ttf that I'm trying to render fonts from by referencing the unicode characters eg. `\uf535'. If I hardcode the text value like this:
> 
> ```dart
> RichText(
>   text: "\uf535"
> )
> ```
> 
> then the icon renders correctly, but doing:
> 
> ```dart
> class Model {
>   String iconString = "\uf535";
> }
> 
> RichText(
>   text: "${model.iconString}"
> )
> ```
> 
> renders the literal string, and no icon/escape formatting works at all. Is there some sort of trick to getting this to work or is something going funky somewhere?



> While this could be more of a SO-type question, [I have posted there already to no useful response](https://stackoverflow.com/questions/57213902/new-line-n-not-being-respected-when-reading-from-sqflite-result) so I'm out of options and this is currently a blocker for many of our production apps that rely on data-driven icons and string formatting.
> 
> I'm trying to use escape characters in both SQFlite data and from other model class variables, neither of which seem to work at all.
> 
> For example, I have a FontAwesome .ttf that I'm trying to render fonts from by referencing the unicode characters eg. `\uf535'. If I hardcode the text value like this:
> 
> ```dart
> RichText(
>   text: "\uf535"
> )
> ```
> 
> then the icon renders correctly, but doing:
> 
> ```dart
> class Model {
>   String iconString = "\uf535";
> }
> 
> RichText(
>   text: "${model.iconString}"
> )
> ```
> 
> renders the literal string, and no icon/escape formatting works at all. Is there some sort of trick to getting this to work or is something going funky somewhere?

I use this:
`String covertStringToUnicode(String content) { String regex = "\\u"; int offset = content.indexOf(regex) + regex.length; while(offset > -1 + regex.length) { int limit = offset + 4; String str = content.substring(offset, limit); if(str != null && str.isNotEmpty){ String code = String.fromCharCode(int.parse(str, radix: 16)); content = content.replaceFirst(str, code, offset); } offset = content.indexOf(regex, limit) + regex.length; } return content.replaceAll(regex, ""); }`
This works perfectly.

**Comment by jamie1192** [comment]
thanks @nxhiep, that's more I less what I ended up doing in the end.

Surely there's got to be a more elegant way to go about this though instead of doing a find/replace on every relevant DB model column though :/
